File: c/examples/connidentity.c
/*
 * connidentity.c — query a device's Identity via class-3 connected
 *                  messaging (Phase 5).  Demonstrates that the
 *                  connected path DOES route to arbitrary backplane
 *                  slots, unlike bp_client_message_send.
 *
 * Usage:
 *   connidentity              # default: slot 2 (L85)
 *   connidentity --slot 1     # explicit
 *
 * SPDX-License-Identifier: MIT
 */
#include <getopt.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bpclient.h"

static void print_id(const uint8_t *resp, uint16_t resp_len) {
    if (resp_len < 4) { printf("  (response too short)\n"); return; }
    uint8_t  reply_svc = resp[0];
    uint8_t  status    = resp[2];
    uint8_t  ext_size  = resp[3];
    printf("  CIP reply: service=0x%02x  general_status=0x%02x  ext_status_sz=%u\n",
           reply_svc, status, ext_size);
    if (status != 0) {
        printf("  CIP error.\n");
        return;
    }
    const uint8_t *body = resp + 4 + ext_size * 2;
    size_t body_len     = resp_len - (4 + ext_size * 2);
    if (body_len < 14) { printf("  (body too short for Identity)\n"); return; }
    uint16_t vendor   = body[0]  | ((uint16_t)body[1]  << 8);
    uint16_t devtype  = body[2]  | ((uint16_t)body[3]  << 8);
    uint16_t prodcode = body[4]  | ((uint16_t)body[5]  << 8);
    uint8_t  major    = body[6];
    uint8_t  minor    = body[7];
    uint32_t serial   = (uint32_t)body[10] | ((uint32_t)body[11] << 8)
                       | ((uint32_t)body[12] << 16) | ((uint32_t)body[13] << 24);
    uint8_t  name_len = body_len > 14 ? body[14] : 0;
    if (15u + name_len > body_len) name_len = (uint8_t)(body_len - 15);
    printf("  Vendor=0x%04x  DevType=0x%04x  ProductCode=0x%04x  fw=%u.%u  serial=0x%08x\n",
           vendor, devtype, prodcode, major, minor, serial);
    if (name_len)
        printf("  Name='%.*s'\n", (int)name_len, body + 15);
}

int main(int argc, char *argv[]) {
    int slot = 2;
    int conn_params = 0x43E8;  /* P2P class-3, variable, 1000 bytes */
    static struct option opts[] = {
        {"slot",        required_argument, 0, 's'},
        {"conn-params", required_argument, 0, 'p'},
        {0,0,0,0}
    };
    int c, idx;
    while ((c = getopt_long(argc, argv, "s:p:", opts, &idx)) != -1) {
        if      (c == 's') slot = atoi(optarg);
        else if (c == 'p') conn_params = (int)strtol(optarg, NULL, 0);
    }

    bp_client_t *cl = NULL;
    if (bp_client_open(&cl) != BP_OK) { fprintf(stderr, "open failed\n"); return 2; }
    bp_client_open_session(cl, NULL);

    /* Build route — port 1 (backplane), link = slot */
    uint8_t epath[2] = { 0x01, (uint8_t)slot };
    bp_conn_spec_t spec = {
        .app_handle   = 1,
        .options      = 0,
        .encoded_path = epath,
        .path_size    = 2,
        .conn_params  = (uint16_t)conn_params,
    };

    int rc;
    uint16_t conn_id = 0, conn_serial = 0;
    rc = bp_client_txrx_open(cl, &spec, &conn_id, &conn_serial);
    printf("[connidentity] slot=%d  conn_params=0x%04x  txrx_open rc=%d (%s)  "
           "conn_id=0x%04x  serial=0x%04x\n",
           slot, (uint16_t)conn_params, rc, bp_strerror(rc), conn_id, conn_serial);

    /* Try TxRxMsg whether or not OpenConn succeeded — some OEM
     * connection models auto-open on first Msg. */
    uint8_t req[]   = { 0x01, 0x02, 0x20, 0x01, 0x24, 0x01 };
    uint8_t resp[256];
    uint16_t resp_len = 0;
    rc = bp_client_txrx_msg(cl, &spec, req, sizeof(req),
                             resp, sizeof(resp), &resp_len);
    printf("[connidentity] txrx_msg rc=%d (%s)  resp_len=%u\n",
           rc, bp_strerror(rc), resp_len);
    if (rc == BP_OK && resp_len > 0) {
        printf("response: ");
        for (uint16_t i = 0; i < resp_len; i++) printf("%02x ", resp[i]);
        printf("\n");
        print_id(resp, resp_len);
    }

    int crc = bp_client_txrx_close(cl, &spec);
    printf("[connidentity] txrx_close rc=%d (%s)\n", crc, bp_strerror(crc));

    bp_client_close(cl);
    return 0;
}

File: python/examples/connidentity.py
#!/usr/bin/env python3
"""connidentity — class-3 connected Identity query.  Mirrors
c/examples/connidentity.c + go/cmd/connidentity CLI + output format.

v0.7.0+: Drives Client.txrx_open / txrx_msg / txrx_close end-to-end
against a real PLC.  The SDK internally builds Large Forward Open +
Forward_Close around the caller's Identity Get_Attributes_All
(svc 0x01) request, all via message_send.

SPDX-License-Identifier: MIT
"""
import argparse
import sys

import bpclient


def parse_hex(s: str) -> bytes:
    out = bytearray()
    i = 0
    while i < len(s):
        while i < len(s) and not _is_hex(s[i]):
            i += 1
        if i + 1 >= len(s) or not _is_hex(s[i + 1]):
            break
        out.append(int(s[i:i + 2], 16))
        i += 2
    return bytes(out)


def _is_hex(c: str) -> bool:
    return c in "0123456789abcdefABCDEF"


def print_id(resp: bytes) -> None:
    if len(resp) < 4:
        print("  (response too short)")
        return
    reply_svc = resp[0]
    status = resp[2]
    ext_size = resp[3]
    print(f"  CIP reply: service=0x{reply_svc:02x}  general_status=0x{status:02x}  "
          f"ext_status_sz={ext_size}")
    if status != 0:
        print("  CIP error.")
        return
    body_start = 4 + ext_size * 2
    if body_start > len(resp):
        return
    body = resp[body_start:]
    if len(body) < 14:
        print("  (body too short for Identity)")
        return
    vendor = body[0] | (body[1] << 8)
    devtype = body[2] | (body[3] << 8)
    prodcode = body[4] | (body[5] << 8)
    major = body[6]
    minor = body[7]
    serial = body[10] | (body[11] << 8) | (body[12] << 16) | (body[13] << 24)
    name_len = body[14] if len(body) > 14 else 0
    if 15 + name_len > len(body):
        name_len = len(body) - 15
    print(f"  Vendor=0x{vendor:04x}  DevType=0x{devtype:04x}  "
          f"ProductCode=0x{prodcode:04x}  fw={major}.{minor}  serial=0x{serial:08x}")
    if name_len > 0:
        print(f"  Name='{body[15:15 + name_len].decode('ascii', 'replace')}'")


def main() -> int:
    ap = argparse.ArgumentParser(prog="connidentity")
    ap.add_argument("--slot", type=int, default=2)
    ap.add_argument("--conn-params", type=lambda x: int(x, 0), default=0,
                    dest="conn_params",
                    help="O→T/T→O size in bytes; 0 = SDK default 4000")
    ap.add_argument("--path", default="")
    ap.add_argument("--app-handle", type=int, default=1, dest="app_handle")
    args = ap.parse_args()

    try:
        client = bpclient.Client()
        client.open()
    except Exception:
        sys.stderr.write("open failed\n")
        return 2
    try:
        client.open_session()
    except Exception:
        pass

    if args.path:
        epath = parse_hex(args.path)
    else:
        epath = bytes([0x01, args.slot])

    spec = bpclient.ConnSpec(
        app_handle=args.app_handle,
        options=0,
        encoded_path=epath,
        path_size=len(epath),
        conn_params=args.conn_params,
    )

    epath_hex = " ".join(f"{b:02x}" for b in epath)
    print(f"[connidentity] app_handle={args.app_handle}  slot={args.slot}  "
          f"conn_params=0x{args.conn_params:04x}  path={epath_hex} ({len(epath)} bytes)")

    try:
        conn_id, conn_serial = client.txrx_open(spec)
        orc = 0
    except Exception as e:
        conn_id = conn_serial = 0
        orc = bpclient.err_code(e)
    print(f"[connidentity] txrx_open rc={orc} (0x{orc & 0xFFFFFFFF:x} {bpclient.strerror(orc)})  "
          f"conn_id=0x{conn_id:04x}  serial=0x{conn_serial:04x}")

    req = bytes([0x01, 0x02, 0x20, 0x01, 0x24, 0x01])
    try:
        resp = client.txrx_msg(spec, req, 256)
        mrc = 0
    except Exception as e:
        resp = b""
        mrc = bpclient.err_code(e)
    print(f"[connidentity] txrx_msg rc={mrc} ({bpclient.strerror(mrc)})  resp_len={len(resp)}")
    if mrc == 0 and resp:
        resp_hex = " ".join(f"{b:02x}" for b in resp)
        print(f"response: {resp_hex} ")
        print_id(resp)

    try:
        client.txrx_close(spec)
        crc = 0
    except Exception as e:
        crc = bpclient.err_code(e)
    print(f"[connidentity] txrx_close rc={crc} ({bpclient.strerror(crc)})")

    client.close()
    ok = (mrc == 0 and len(resp) >= 4 and resp[0] == 0x81 and resp[2] == 0x00)
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())
